source/language/openpulse.rst

```diff
@@ -0,0 +1,742 @@
+- Clearly defined relationship between pulses, channels and the phases of compilation to hardware resources.
+- Use of multiple oscillators on a single channel at the same time (see :ref:`Geo Gate`):
+
+Openpulse provides a flexible programming model that should extend to many quantum control schemes and hardware.
+At the core of the OpenPulse grammar are the concepts of ``signals`` and ``channels``. Many of these concepts
+are inspired by the signal processing language <Faust `https://faust.grame.fr/`>_.
+``Signals`` define a discrete set of samples (or waveform) which are then played or captured via a
+a ``channel`` on the target device. A channel is a software abstraction which
+allows the programmer to be agnostic to complexities of the device's underlying pulse generation
+hardware. It is the responsibility of the target device's compiler to map ``signals``
+to the applied channels in target hardware.
+
+**NOTE: We assume we have arrays defined with a [...] symtax and complex[size] types.**
+
+Signals
+-------
+
+A ``signal`` is a generalized concept of a ``pulse``. A signal is a discrete, time-dependent function
+:math:`s(t): ℤ->ℂ`. The input time is an integer representing the number of elapsed samples. It can
+be viewed as the "index" into the signal. In hardware, the time :math:`t_h`, is given by
+:math:`t_h=t*dt`, where ``dt`` is the time elapsed per sample. Signals have a max unit-norm
+(signal processes that emit a signal outside of this range clip).
+
+All signals have a ``.duration`` property giving the length of the signal.
+
+.. code-block:: c
+
+    signal sig = [0.1, 1.+1I, 0.1 - 0.2I];
+    int[32] x = sig.duration + 1;  // x is 4
+
+Signals are produced from two fundamental types, ``envelopes`` and ``carriers``.
+
+Envelopes
+~~~~~~~~~
+
+Envelopes are discrete waveforms :math:`f(t), ℤ->ℂ`. Examples include a waveform described as
+an array of complex samples or a parametric function supported directly by the hardware
+such as a Gaussian waveform.
+
+Arrays initialization is done as in OpenQASM.
+
+.. code-block:: c
+
+    envelope env = [0.1, 1.+1I, 0.1 - 0.2I, ...];
+
+Parametric pulses are initialized by assigning the result of``envelope``'s to the result of a (kernel) function call.
+
+.. code-block:: c
+
+  envelope env = gaussian(1.+1I, 1024dt, 128dt);
+
+The following parametric functions are currently defined
+
+.. code-block:: c
+
+    kernel gaussian(complex[size] amp, length duration, length sigma) -> envelope;
+    kernel gaussian_square(complex[size] amp, length duration, length sigma, length square_width) -> envelope;
+    kernel drag(complex[size] amp, length duration, length sigma, float[size] beta) -> envelope;
+    kernel constant(complex[size] amp, length duration) -> envelope;
+
+Carriers
+~~~~~~~~
+
+Carriers represent stateful modulation signals in hardware such as those used to implement the <Virtual Z-gate `https://arxiv.org/abs/1612.00858`>_.
```

Comment:
I know I'm fighting a loosing battle but it kills me that was known and used for decades before arXiv:1612.00858. E.g. in ion traps from 2008 [Randomized benchmarking of quantum gates](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.77.012307) this was just an offhand comment:

> Pulses involving z-axis rotations were implemented by programmed phase changes of one of the Raman beams. This changes the phase of the rotating reference frame and is equivalent to the desired z-axis rotation.